### DISCUSSION TITLE
Dropping support for some .NET versions

### DISCUSSION THREAD

**Discussion by bruno-garcia** [getsentry/sentry-dotnet#2776] [General]
Sentry's core package and some integrations such as ASP.NET Core support a LOT of framework versions. Many not supported by Microsoft and the .NET team anymore. Since this package was created back in 2018, we have never dropped support to any framework version.

We're working the 4.0 release now and the time to drop some of these old frameworks came up. 

But why are you doing that Sentry? When it's part of [your philosophy that _backward compatibility is king_](https://develop.sentry.dev/sdk/philosophy/#compatibility-is-king):

Because it's really hard. Every time we make a PR, we get CI to run about **11.700 tests**. That's a lot of tests for an SDK. 

The Sentry core package alone has **1442 tests**. But it gets run several times,  **8888** to be precise. That's because it's one run for each target framework that we support.

Additionally, we run tests (when applicable) on: **Windows, macOS, Linux, iOS and Android**. 

Not only CI times, but the complexity of building new features, bug fixes, etc in all of those TFMs is really a lot of work.
We want to spend time building new things for you all.

### Dropped framework versions

1. .NET Core 3.0

This has been out of support _March 3, 2020_
We don't intend to replace it with .NET Core 3.1, because that's also out of support, since _December 13, 2022_

2. .NET 5

Out of support since _May 10, 2022_. This was a _Current_ release and even .NET Core 3.1 outlived it.

3. .NET Framework 4.6.1 -> 4.6.2

.NET Framework 4.6.1 was announced in _Nov 30, 2015_. And went out of support over a year ago, on _Apr 26, 2022_.
Looking at support from some popular libraries, .NET Framework 4.6.2 seems now be the baseline (this is the lowest supported version). For example, [`Microsoft.Extensions.Logging` for example targets 4.6.2.](https://www.nuget.org/packages/Microsoft.Extensions.Http/). 

.NET Framework 4.6.2 was announced in _August 2nd, 2016_ and is going out of support in 2 months already. 

4. .NET 6 Android & iOS

In order to support .NET 8 for Android and iOS, we're left with the choice of either bundling the old builds on the new SDK, or simply drop support .NET 6 Mobile (`net6.0-android` and `net6.0-ios`). That's because Microsoft dropped support **for building** .NET 6 apps on mobile with the .NET 8 SDK. But since it's extremely likely (lots of bugs on those early releases!!) you'll be upgrading a mobile app on .NET to 7 (and likely 8 as soon as it's out in 2 weeks), we decided with dropping support. **We're replacing it with support for .NET 7 instead.**

More context here:

* https://github.com/getsentry/sentry-dotnet/issues/2623

5. .NET Standard 2.1

This isn't decided today as much as a question: Do you target ns2.1 anywhere? If so, why? We feel we can keep ns2.1 as a baseline to have high interoperability (e.g: Unity) and 2.1 was just a .NET Core 3.1 bridge release and won't help us anymore here.

### What packages get affected?

These frameworks would be dropped not only on the core `Sentry` package, but all integrations such as `Sentry.AspNetCore` and `Sentry.Serilog`.

### What versions are left?

1. .NET Standard 2.0 where we currently have it. This is used by our [Unity](https://github.com/getsentry/sentry-unity/) and [Xamarin](https://github.com/getsentry/sentry-xamarin) SDKs anyway, even Sentry needs it.
2. .NET Framework 4.6.2
3. .NET 6 and later (7, 8) where applicable

Would love to hear YOUR thoughts: What do you need?

**Comment by SimonCropp** [comment]
unfortunately keeping support for .NET Standard 2.0 implies support for many runtimes
![image](https://github.com/getsentry/sentry-dotnet/assets/122666/ff907e5a-059f-4310-9321-598dbb7961f3)

If u want to "keep NET Standard 2.0 for only unity" i think u need some msbuild that throws a build error if the project target is not .NET Standard 2.0.

**Reply by bruno-garcia** [reply]
That's one of the reasons we'd like to keep it. To have that 'catch all' support through ns2.0 and ultimate backward compatibility. But some functionality might not be available there, and it includes some dependencies. We use that on Xamarin too since the old Xamarin mobile TFMs pull ns2.0 from Sentry

**Comment by SimonCropp** [comment]
@bruno-garcia one of those awesome pie charts with "what runtimes sentry users are running" would be helpful

**Reply by bruno-garcia** [reply]
~10% of events come from unsupported runtimes

**Reply by SimonCropp** [reply]
and how many of those are using a recent version of the sentry nuget?

**Reply by bruno-garcia** [reply]
it's folds down slowly, if I look at the latest version it's less than that. If I look at versions released in the last 3 months, it more than that. So folks update Sentry but not their runtime

**Comment by SimonCropp** [comment]
one other suggestion i have:  u could potentially do a release now with a msbuild warning that detects targeting of any of the planned obsolete ones.  and links to this GH issue

**Reply by bruno-garcia** [reply]
That's a good idea! Thanks Simon

**Comment by Tyrrrz** [comment]
.NET 4.6.2+ and .NET 6.0+ (or better, .NET 8.0+) seem like the two most reasonable targets to support.

In my opinion, .NET Standard has become obsolete ever since .NET 5, when the rift in supported APIs started growing and .NET Standard was no longer expanded. I see no point in targeting .NET Standard anymore, unless it's the only target in the project (i.e. it doesn't require any of the newer APIs and doesn't use any platform-specific APIs).

From my personal perspective, since I'd be using Sentry in apps, upgrading to the latest runtime is really easy, so I only need it to target the latest .NET version available.

**Reply by Tyrrrz** [reply]
Might be relevant, I was planning to do .NET 5+ back in 2021 but ultimately decided that the breaking changes were not worth it back then. You might find the comments useful perhaps:
- https://github.com/Tyrrrz/CliWrap/issues/131
- https://github.com/Tyrrrz/YoutubeExplode/issues/609
- https://github.com/Tyrrrz/CliFx/issues/116

**Reply by Tyrrrz** [reply]
At the very least, I would drop .NET Standard and .NET < 5 from my projects now. I just see no reason to do it without bundling with some other breaking changes along the way.

**Reply by bruno-garcia** [reply]
If I was maintaining a library on my spare time I'd definitely drop .NET < 5 and Standard. But for Sentry, we want to allow the most number of possible use cases. So we're OK keeping some old stuff around, but definitely don't need the amount of TFMs we accumulated.

Thanks for the links. Helps, like:

> I cant upgrade to net5.0 hince I am using mono
> I cant use dotnet core because it is not supported by powerpc

> Please keep the implementation in a way that supports .NETSTANDARD 2.0
> Therefore UWP and .NET Framework users can use it too.

Useful comments to learn about some use cases of folks stuck on older frameworks.

Xamarin support ends on _May 1, 2024_ but we're not going to stop supporting it immediately. So ns2.0 has to stay for another year at least I think.

**Reply by cartermp** [reply]
Hmmm. Those circumstances represent some pretty serious edge cases, though. How much money is attached to each of those? Perhaps Sentry has some product telemetry to tell, because if it's just like 3 people on powerpc spending 30 bucks a month, then if I were running this project I'd say "too bad, so sad" 🙂

**Reply by cartermp** [reply]
FWIW this is what we did at Microsoft for F#. Especially for compiler+tools support, older frameworks are expensive to maintain. Once I was able to get an accurate range for the actual money at stake for some of this older stuff, and found that it was basically marginal, the decision was incredibly easy.

**Reply by bruno-garcia** [reply]
We have numbers actually (not sure about powerpc, agree that's super edge case). But how many are on outdated runtimes and which ones and which versions. It's no insignificant. It wasn't a lot of work to add the new framework support from the get-go, but it has become a burden recently, reason why we're opening this discussion.

Wdyt about the plan? Keep net462, ns2.0 and net6.0+? Prob next year we can cut off ns2.0.

**Comment by bruno-garcia** [comment]
I verified and the number of projects on ASP.NET Core on .NET Framework is quite low. So adding it to the list:

Removing .NET Standard 2.0 from `Sentry.AspNetCore`

**Comment by bruno-garcia** [comment]
4.0.0 is live with the new changes.
Examples:

[Sentry](https://www.nuget.org/packages/Sentry/4.0.0)
![image](https://github.com/getsentry/sentry-dotnet/assets/1633368/f87e6403-9379-49f7-bbe0-633bcb614df1)
[Sentry.AspNetCore](https://www.nuget.org/packages/Sentry.AspNetCore)
![image](https://github.com/getsentry/sentry-dotnet/assets/1633368/c60ba96d-6c12-43e7-99ec-1305ea7d8bd5)
